scripts/cedict.js

```diff
@@ -98,241 +75,212 @@ const parseClassifiers = (classifiersString) => {
-    index: 0
-  }
-
-  if (/·/.test(parts[1])) {
-    // This is a compound name (e.g. a western name consisting of first and last names)
-    entryTemplate.type = 'compound name'
-
-    let subparts = parts[1].match(/(\S+)·(\S+)/)
-    if (subparts) {
-      entryTemplate.traditionalHeadword = ''
-      entryTemplate.traditionalHeadwordParts = {
-        firstName: subparts[1],
-        lastName: subparts[2]
-      }
-    } else {
-      console.error(`Cannot parse a compound name in a traditional headword in line ${counter}: ${line}`)
-    }
-
-    subparts = parts[2].match(/(\S+)·(\S+)/)
-    if (subparts) {
-      entryTemplate.simplifiedHeadword = ''
-      entryTemplate.simplifiedHeadwordParts = {
-        firstName: subparts[1],
-        lastName: subparts[2]
-      }
-    } else {
-      console.error(`Cannot parse a compound name in a traditional headword in line ${counter}: ${line}`)
-    }
-
-    subparts = parts[3].match(/(.+)\s·\s(.+)/)
-    if (subparts) {
-      entryTemplate.pinyin = ''
-      entryTemplate.pinyinParts = {
-        firstName: subparts[1],
-        lastName: subparts[2]
-      }
-    } else {
-      console.error(`Cannot parse a compound name in a traditional headword in line ${counter}: ${line}`)
-    }
-  } else if (/.+，.+/.test(parts[1])) {
-    // This is a proverb
-    entryTemplate.type = 'proverb'
-    entryTemplate.traditionalHeadword = parts[1]
-    entryTemplate.simplifiedHeadword = parts[2]
-    entryTemplate.pinyin = parts[3]
-  } else {
-    // This is a regular dictionary entry
-    entryTemplate.type = 'not specified'
-    entryTemplate.traditionalHeadword = parts[1]
-    entryTemplate.simplifiedHeadword = parts[2]
-    entryTemplate.pinyin = parts[3]
-  }
-
-  let parsedDescriptions = [] // eslint-disable-line prefer-const
-  parseDescription(parts[4], parsedDescriptions)
-  parsedDescriptions.forEach((desc) => {
-    // If there are multiple descriptions then this word is a homonym and shall produce multiple dictionary entries
-    let entry = Object.assign({}, entryTemplate) // eslint-disable-line prefer-const
-    entry.index = counter
-    entry.descriptions = desc.descriptions.split('/')
-    if (desc.hasOwnProperty('classifier')) { // eslint-disable-line no-prototype-builtins
-      entry.classifier = parseClassifiers(desc.classifier)
-    }
-    dictData.entries.push(entry)
-    sizeData.push({
-      index: counter,
-      size: encoder.encode(JSON.stringify(entry, ...stringifyOptions)).length
-    })
-    counter++
-  })
-})
+/**
+ * Parses CEDICT metadata.
+ *
+ * @param {string} metadataStr - A string containing CEDICT metadata.
+ * @returns {object} A CEDICT metadata object.
+ */
+const parseCedictMeta = (metadataStr) => {
+  let metaData = {} // eslint-disable-line prefer-const
 
-rlInterface.on('close', () => {
   // Parse metadata
   let match = metadataStr.match(/^#\s(.+)/)
   if (match !== null) {
-    dictData.metadata.name = match[1]
+    metaData.name = match[1]
```

Comment:
I could be wrong here, but are you the author of the metadata?
Why should be it parsed from such an unstructed source with the help of regexep?

I believe that using regexp could have the following disadvantages:
   - it is a heavy calculation process
   - regexp standards could be changed upon different updates (it is not standartized across all platforms)
   - it could be easily broken with small changes to the parsed text and it could be difficult to find the problem in such a case (especcially when it should be done not by the author or in some longer time in future)

May be there is a way to structure metadata nad get the info from the structure but not raw text?

Reply:
This parses a CEDICT metadata to convert it into a JSON-friendly format. I agree that regexps are mind-boggling sometimes, but they have some important advantages too. It takes longer to calculate a regexp usually, but you can get several capturing groups at once. Doing the same with regular string search functions would take much more code. It will be eventually harder to read, on my opinion. Speed is generally not an issue here as the whole process takes just a couple of seconds (although I do not think they are that much slower to cause any problems).

I agree that regexp can differ between platforms, but in our case they are executing in node.js environment only. We had to update these scripts after data changes anyway. I think it's a process we cannot avoid.

I also agree that problems with regexps can be harder to find. That's why I tried to put error messages indicating a faulty place everywhere.

What I like about regexps is that they are very concise and allows to do very complex parsing with just one line of code. Also they are kind of cross tool technology (but that's not a point here). But I also agree they have their portion of issues. I just think in this case advantages probably outweighs drawbacks, especially in more complex parsing expressions as `/(\S+) (\S+) \[(.+)] \/(.+)\//`. But that's just my opinion, of course.